### PR TITLE
Zerodha ProductType customization for Fetching Positions&holdings and also while placing orders  

### PR DESCRIPTION
**Pull request by rjra2611**
#### Description

- zerodha-product-type no more a compulsory argument in config
- User support to provide  zerodha-product-type per order
- Fetch holdings as per zerodha-product-type in config otherwise all
- Use withlock instead of lock/unlock

#### Related Issue

- Related to https://github.com/QuantConnect/Lean/pull/5710
#### Motivation and Context

Handles the use case where Alphas' use long term holdings as well as intraday positions

#### Requires Documentation Change

- N/A

#### How Has This Been Tested?

Only on windows Machine

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Brokerages/Zerodha/ZerodhaBrokerage.cs [hunk 1]
```diff
@@ -438,85 +413,92 @@ public override void Disconnect()
...
+                    {
+                        order.BrokerId.Add(brokerId);
+                        CachedOrderIDs.TryAdd(order.Id, order);
+                    }
+
+                    // Generate submitted event
+                    OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, orderFee, "Zerodha Order Event") { Status = OrderStatus.Submitted });
+                    Log.Trace($"Order submitted successfully - OrderId: {order.Id}");
+
+                    submitted = true;
```
Comment (Martin-Molinero) [inline Brokerages/Zerodha/ZerodhaBrokerage.cs:492]: I think there is no need to set `submitted` to false, it's the default value? only need to set it on true/success, should directly return here? If it doesn't return if will emit the failure bellow? Similar comment for the other methods 👍

File: Brokerages/Zerodha/ZerodhaBrokerage.cs [hunk 2]
```diff
@@ -357,32 +358,6 @@ private void OnOrderUpdate(Messages.Order orderUpdate)
...
-        /// <summary>
-        /// Unlock stream and process all backed up messages.
-        /// </summary>
-        public void UnlockStream()
-        {
-            Log.Trace("ZerodhaBrokerage.Messaging.UnlockStream(): Processing Backlog...");
-            while (_messageBuffer.Any())
-            {
-                MessageData e;
-                _messageBuffer.TryDequeue(out e);
```
Comment (Martin-Molinero) [inline Brokerages/Zerodha/ZerodhaBrokerage.cs:378]: `_messageBuffer` can be removed now? `LastHeartbeatUtcTime, ApiSecret, ChannelList` are not used, should we remove or used them?

File: Common/Orders/ZerodhaOrderProperties.cs [hunk 3]
```diff
@@ -25,6 +25,11 @@ namespace QuantConnect.Orders
     public class ZerodhaOrderProperties : OrderProperties
     {
 
+        /// <summary>
+        /// Kite product type
+        /// </summary>
+        public readonly string ProductType;
```
Comment (Martin-Molinero) [inline Common/Orders/ZerodhaOrderProperties.cs:31]: Let's make it a property -> `{get;}`

File: Common/Orders/ZerodhaOrderProperties.cs [hunk 4]
```diff
@@ -33,6 +38,37 @@ public ZerodhaOrderProperties(string exchange) : base(exchange)
         {
         }
 
+        /// <summary>
+        /// Initialize a new OrderProperties for <see cref="ZerodhaOrderProperties"/>
+        /// </summary>
+        /// <param name="exchange">Exchange value, nse/bse etc</param>
+        /// <param name="productType">Product type</param>
+        public ZerodhaOrderProperties(string exchange,KiteProductType productType) : this(exchange)
```
Comment (Martin-Molinero) [inline Common/Orders/ZerodhaOrderProperties.cs:46]: nit - missing space between `exchange,KiteProductType`

File: Brokerages/Zerodha/ZerodhaBrokerage.cs [hunk 5]
```diff
@@ -826,63 +816,62 @@ private OrderStatus ConvertOrderStatus(Messages.Order orderDetails)
...
-        /// Gets all open positions
+        /// Gets all open postions and account holdings
         /// </summary>
         /// <returns></returns>
         public override List<Holding> GetAccountHoldings()
         {
             var holdingsList = new List<Holding>();
-            if (_zerodhaProductType.ToUpperInvariant() == KiteProductType.MIS.ToString().ToUpperInvariant() || _zerodhaProductType.ToUpperInvariant() == KiteProductType.NRML.ToString().ToUpperInvariant())
+            // get MIS and NRML Positions
+            if (string.IsNullOrEmpty(_zerodhaProductType) || _zerodhaProductType.ToUpperInvariant() == KiteProductType.MIS.ToString().ToUpperInvariant() || _zerodhaProductType.ToUpperInvariant() == KiteProductType.NRML.ToString().ToUpperInvariant())
```
Comment (Martin-Molinero) [inline Brokerages/Zerodha/ZerodhaBrokerage.cs:826]: nit - line is a bit long? Can perform `_zerodhaProductType.ToUpperInvariant()` once I think?

File: Brokerages/Zerodha/ZerodhaBrokerage.cs [hunk 6]
```diff
@@ -593,97 +575,105 @@ private static string ConvertOrderType(OrderType orderType)
...
-            {
+                    submitted = false;
+                }
+                if (orderProperties.ProductType != null)
+                {
+                    zerodhaProductType = orderProperties.ProductType;
+                }
+                else if (string.IsNullOrEmpty(zerodhaProductType))
+                {
+                    throw new ArgumentException("Please set ProductType in config or provide a value in DefaultOrderProperties");
```
Comment (Martin-Molinero) [inline Brokerages/Zerodha/ZerodhaBrokerage.cs:604]: Cool! 👍

**Review by Martin-Molinero — COMMENTED** [review]
Nice improvement! 👍 leaving a couple of minor requests

File: Brokerages/Zerodha/ZerodhaBrokerage.cs (hunk 2 above)
Comment (rjra2611) [reply Brokerages/Zerodha/ZerodhaBrokerage.cs:378]: `ApiSecret` is being used to get `Kite `object to deal with zerodha api
Rest all we can remove

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**